### PR TITLE
Implement parametric profile.

### PR DESCRIPTION
**Pull request by ikeough**
BACKGROUND:
See #696 

DESCRIPTION:
- Adds a new type, `ParametricProfile` which derives from `Profile` and enables the user to specify the profile's vertex locations using expressions for X, and Y.
- Refactors profile servers to implement the `IProfileFactory` interface. "Old" profile servers derive from a base class `ProfileFactoryBase`, while all new profile servers derive from `ParametricProfileFactory`.
- Adds C, L, WT, W, MC, HSS, and ST sections implemented as section catalogues and parametric profile factories.

TESTING:
- See `ParametricProfileTests`
  
FUTURE WORK:
- The original `WideFlangeProfile` is deprecated and should be removed.
- The original `HSSPipeProfile` should be converted to a parametric profile type and the original should be deprecated.

REQUIRED:
- [x] All changes are up to date in `CHANGELOG.md`.

image
image
image
image
image
image

---
This change is [Reviewable](https://reviewable.io/reviews/hypar-io/elements/697)

### CODE REVIEW

File: Elements/src/Geometry/Profiles/ParametricProfile.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                throw new ArgumentException("The parametric profile could not be created. No translation expressions were provided.");
+            }
+
+            Data = data;
+
+            // These references are required to get dynamic to work.
+            _refs = new List<MetadataReference>{
+                                    MetadataReference.CreateFromFile(typeof(Microsoft.CSharp.RuntimeBinder.RuntimeBinderException).GetTypeInfo().Assembly.Location),
+                                    MetadataReference.CreateFromFile(typeof(System.Runtime.CompilerServices.DynamicAttribute).GetTypeInfo().Assembly.Location)};
+            Perimeter = CreatePolygonFromExpressions(data.PerimeterVectorExpressions, data.PropertyValues, _refs).Result;
```
Comment (radumg) [inline Elements/src/Geometry/Profiles/ParametricProfile.cs:45]: using `.Result` in ctors can/will likely result in deadlocks
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#constructors

**Review by andrewheumann — CHANGES_REQUESTED** [review]
> Reviewed 2 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/697)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @ikeough and @radumg)

___
*[Elements/src/Geometry/Profile.cs, line 26 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-Mnjknwv2mcCiDaftnXp:-Mnjknww36zxsQYFHALE:b-haxd4x) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profile.cs#L26)):*
> ```Smalltalk
>         /// and is not intended to be used directly.
>         /// </summary>
>         public Profile() { }
> ```

should this be `internal`, rather than `public`?
___
*[Elements/src/Geometry/Profile.cs, line 266 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-MnjixpA-gAj1yjeB0LF:-MnjixpA-gAj1yjeB0LG:bm73xh0) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profile.cs#L266)):*
> ```Smalltalk
>             // profile which defines its own perimeter and void logic
>             // during construction.
>             if (Perimeter == null || Voids == null)
> ```

are we 100% sure that profiles can't ever have null Voids in non-parametric profile cases? I think this is still possible See the `Profile(Polygon perimeter, Polygon @void)` constructor for one way to construct profiles that doesn't involve checking for a null `Voids` array. (We should probably just fix that instead of this, but maybe a `if this is ParametricProfile` or something would be prudent too?)
___
*[Elements/src/Geometry/Profiles/ParametricProfile.cs, line 23 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-MnjkwD32NuwZXDl1rO8:-MnjkwD32NuwZXDl1rO9:bm40a6p) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profiles/ParametricProfile.cs#L23)):*
> ```Smalltalk
>         /// The data which defines the profile.
>         /// </summary>
>         public ParametericProfileData Data { get; set; }
> ```

Parameteric  => Parametric

**Review by ikeough — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/697)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @andrewheumann and @radumg)

___
*[Elements/src/Geometry/Profile.cs, line 26 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-Mnjknwv2mcCiDaftnXp:-Mnq7cIK8Wj0ePfuSpDR:bb6un1j) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profile.cs#L26)):*
<details><summary><i>Previously, andrewheumann wrote…</i></summary><blockquote>

should this be `internal`, rather than `public`?
</blockquote></details>

Done
___
*[Elements/src/Geometry/Profile.cs, line 266 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-MnjixpA-gAj1yjeB0LF:-Mnq7jXzA_JmCGip0kpn:b3nnnqt) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profile.cs#L266)):*
<details><summary><i>Previously, andrewheumann wrote…</i></summary><blockquote>

are we 100% sure that profiles can't ever have null Voids in non-parametric profile cases? I think this is still possible See the `Profile(Polygon perimeter, Polygon @void)` constructor for one way to construct profiles that doesn't involve checking for a null `Voids` array. (We should probably just fix that instead of this, but maybe a `if this is ParametricProfile` or something would be prudent too?)
</blockquote></details>

In any case where voids can be null, this would be necessary. Are you asking if I should just remove the comment? 
___
*[Elements/src/Geometry/Profiles/ParametricProfile.cs, line 23 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-MnjkwD32NuwZXDl1rO8:-Mnq8KGz7cTRVP5HCwKZ:b-896fix) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profiles/ParametricProfile.cs#L23)):*
<details><summary><i>Previously, andrewheumann wrote…</i></summary><blockquote>

Parameteric  => Parametric
</blockquote></details>

Done.

**Review by andrewheumann — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/697)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @ikeough and @radumg)

___
*[Elements/src/Geometry/Profile.cs, line 266 at r2](https://reviewable.io/reviews/hypar-io/elements/697#-MnjixpA-gAj1yjeB0LF:-Mnr5VIi0NxnSypUeZFS:blswq4s) ([raw file](https://github.com/hypar-io/elements/blob/21a28b9b2b794c44c81dd063bd8a2cf1cadc39de/Elements/src/Geometry/Profile.cs#L266)):*
<details><summary><i>Previously, ikeough (Ian Keough) wrote…</i></summary><blockquote>

In any case where voids can be null, this would be necessary. Are you asking if I should just remove the comment? 
</blockquote></details>

oh, that makes sense — was thrown off by the comment. disregard.

**Review by andrewheumann — COMMENTED** [review]
> Reviewed 1 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/697)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @ikeough and @radumg)

**Pull request merged**